Standardise scores to `long` type

maromalo opened the pull request:
Addresses #17783 and #20298.

Score tests are now off by 1 point, I guess it's expected since results are floored instead of rounded. Not sure if to adjust the tests or the scoring.

maromalo committed:
Make score type consistent

maromalo committed:
quality shenanigans

smoogipoo commented on osu.Game/Rulesets/Scoring/ScoreProcessor.cs:
```diff
@@ -333,21 +333,21 @@ public double ComputeScore(ScoringMode mode, ScoringValues current, ScoringValue
                     double accuracyScore = accuracyPortion * accuracyRatio;
                     double comboScore = comboPortion * comboRatio;
-                    return (max_score * (accuracyScore + comboScore) + bonusScore) * scoreMultiplier;
+                    return (long)((max_score * (accuracyScore + comboScore) + bonusScore) * scoreMultiplier);
```
Fairly certain this shouldn't be rounded when called recursively via the `Classic` path.

smoogipoo commented on osu.Game/Rulesets/Scoring/ScoreProcessor.cs:
```diff
@@ -417,7 +417,7 @@ public virtual void PopulateScore(ScoreInfo score)
 
             // Populate total score after everything else.
-            score.TotalScore = (long)Math.Round(ComputeScore(ScoringMode.Standardised, score));
+            score.TotalScore = ComputeScore(ScoringMode.Standardised, score);
```
I'm concerned about all these places using round, and the new code using truncation (floor). On osu-stable there's an issue where mania scores could get to 999999 with 100% acc that's with a jank conditional. I believe we still want rounding to avoid such an issue appearing again here.

smoogipoo reviewed (changes requested):
Rest looks fine

maromalo commented on osu.Game/Rulesets/Scoring/ScoreProcessor.cs:
```diff
@@ -333,21 +333,21 @@ public double ComputeScore(ScoringMode mode, ScoringValues current, ScoringValue
                 case ScoringMode.Standardised:
                     double accuracyScore = accuracyPortion * accuracyRatio;
                     double comboScore = comboPortion * comboRatio;
-                    return (max_score * (accuracyScore + comboScore) + bonusScore) * scoreMultiplier;
```
Turns out this and `ScoringMode.Classic`'s return value need rounding (for example at `TestFourVariousResultsOneMiss` it returns `387499.99999999994` instead of `387500`). Before the type change they only got rounded in `PopulateScore()`.

smoogipoo replied:
I'm still not sure about your change here. I meant that you'd want this to _not_ be rounded for `Classic`, so the implementation would be something like:

```
long GetScore()
{
    if (Standardised)
        return (long)Math.Round(getScoreRaw(...));
    else
        return (long)Math.Round(getScoreRaw(...) * classic_stuff);

    double getScoreRaw(...)
    {
        // Raw standardised score.
    }
}
```

maromalo replied:
Oh, sorry. I'll be pushing a fix in a few minutes.

maromalo committed:
Fix score rounding issue

maromalo committed:
Add computeRawScore()

maromalo committed:
Merge branch 'master' into long-standard

smoogipoo committed:
Compute raw score in-line

Saves on an allocation (though it seems to only be a ValueType so it
shouldn't be a big deal) and a tail-call.

smoogipoo committed:
Merge branch 'master' into long-standard

smoogipoo reviewed (approved):
LGTM. I haven't tested osu-web/osu-queue-score-statistics/osu-elastic-indexer, however `SoloScoreInfo` was already an `int` (used by both score-statistics and elastic-indexer) and elastic-indexer defines it as an integral type in the schema. So I have no reason to believe this will break anything.
